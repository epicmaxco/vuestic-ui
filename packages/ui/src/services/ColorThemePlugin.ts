import Vue from 'vue'
import { makeContextablePropsMixin } from '../components/context-test/context-provide/ContextPlugin'
import { Component } from 'vue-property-decorator'

declare module 'vue/types/vue' {
  interface Vue {
    // @deprecated, use getColor instead
    $themes: Record<string, string> | undefined;
  }
}

// Most default color - fallback when nothing else is found.
const DEFAULT_COLOR = '#000000'

const defaultOptions = Vue.observable({
  themes: {
    primary: '#23e066',
    secondary: '#002c85',
    success: '#40e583',
    info: '#2c82e0',
    danger: '#e34b4a',
    warning: '#ffc200',
    gray: '#b4b6b9',
    dark: '#34495e',
  },
})

export const THEME_NAMES = {
  DEFAULT: 'DEFAULT',
  CORPORATE: 'CORPORATE',
}

export const COLOR_THEMES = [
  {
    name: THEME_NAMES.DEFAULT,
    themes: {
      primary: '#40e583',
      secondary: '#002c85',
      success: '#40e583',
      info: '#2c82e0',
      danger: '#e34b4a',
      warning: '#ffc200',
      gray: '#babfc2',
      dark: '#34495e',
    },
  },
  {
    name: THEME_NAMES.CORPORATE,
    themes: {
      primary: '#6c7fee',
      secondary: '#6e7ff1',
      success: '#8ddc88',
      info: '#71baff',
      danger: '#f8706d',
      warning: '#ffd652',
      gray: '#8396a5',
      dark: '#34495e',
    },
  },
]

export const getDefaultOptions = () => defaultOptions

export const ColorThemePlugin = {
  install (Vue: Vue, options?: { themes?: Record<string, string> }) {
    if (options && options.themes) {
      defaultOptions.themes = { ...defaultOptions.themes, ...options.themes }
    }

    // @ts-ignore
    Vue.prototype.$themes = defaultOptions.themes
  },
}

/**
 * Check if color is valid css color
 * Taken from https://stackoverflow.com/a/56266358/5783475
 * @param strColor
 */
export const isCssColor = (strColor: string): boolean => {
  const s = new Option().style
  s.color = strColor
  return s.color !== ''
}

export const getColor = ($vm: Vue, prop: string, defaultColor: string = DEFAULT_COLOR): string => {
  if ($vm.$themes && $vm.$themes[prop]) {
    return $vm.$themes[prop]
  }

  if (isCssColor(prop)) {
    return prop
  }

  return defaultColor
}

@Component({
  mixins: [
    makeContextablePropsMixin({
      color: {
        type: String,
      },
    }),
  ] as any,
})
export class ColorThemeMixin extends Vue {
  get colorComputed () {
    return getColor(this, (this as any).c_color)
  }

  computeColor (prop: string, defaultColor?: string) {
    return getColor(this, prop, defaultColor)
  }

  setTheme (theme: Record<string, string>) {
    Object.assign(this.$themes, theme)
  }

  created () {
    (this as any).hasColorThemeMixin = true
  }
}
