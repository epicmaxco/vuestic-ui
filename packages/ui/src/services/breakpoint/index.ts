import { BreakpointConfig, ThresholdsConfig } from './types'

export const vaBreakpointSymbol = Symbol('vaBreakpoint')

export const defaultThresholds: ThresholdsConfig = {
  xs: 0,
  sm: 640,
  md: 1024,
  lg: 1440,
  xl: 1920,
}

export const getBreakpointDefaultConfig = (): BreakpointConfig => ({
  enabled: true,
  bodyClass: true,
  thresholds: defaultThresholds,
})

export * from './types'
