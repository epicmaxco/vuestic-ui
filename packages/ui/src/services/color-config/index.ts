export * from './types'
export * from './composables/useTheme'
