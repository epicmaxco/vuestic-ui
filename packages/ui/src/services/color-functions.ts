import { ColorTranslator } from 'colortranslator'
import { HSLObject, ColorInput } from 'colortranslator/dist/@types'

export const colorToRgba = (color: ColorInput, opacity: number) => {
  return new ColorTranslator(color).setA(opacity).RGBA
}

export const getBoxShadowColor = (color: ColorInput) => {
  return new ColorTranslator(color).setA(0.4).RGBA
}

export const getHoverColor = (color: ColorInput) => {
  return new ColorTranslator(color).setA(0.2).RGBA
}

export const getFocusColor = (color: ColorInput) => {
  return new ColorTranslator(color).setA(0.3).RGBA
}

export const shiftHslColor = (color: ColorInput, offset: { h?: number; s?: number; l?: number; a?: number }) => {
  const result = new ColorTranslator(color)

  if (offset.h) {
    result.setH(result.H + offset.h)
  }

  if (offset.s) {
    result.setS(result.S + offset.s)
  }

  if (offset.l) {
    result.setL(result.L + offset.l)
  }

  if (offset.a) {
    result.setA(result.A + offset.a)
  }

  return result.HSL
}

/**
 * This method guesses second color for gradient based on first one.
 * Calculation method should differ for different colors, so we split color circle
 * in zones.
 *
 * @param color
 */
export const shiftGradientColor = (color: ColorInput): string => {
  const newColor = ColorTranslator.toHSL(color, false) as HSLObject

  // Gray
  if (newColor.s < 10) {
    return shiftHslColor(newColor, { h: 2, s: 5, l: 10 })
  }
  // Unsaturated (too pale, gray-ish, monotone)
  // For unsaturated color, hue change does nothing.
  if (newColor.s < 30) {
    return shiftHslColor(newColor, { s: -14, l: 11 })
  }
  // Red
  if ((newColor.h >= 0 && newColor.h < 44) || (newColor.h >= 285)) {
    return shiftHslColor(newColor, { h: 11, s: 27, l: 8 })
  }
  // Yellow
  if (newColor.h >= 44 && newColor.h < 85) {
    return shiftHslColor(newColor, { h: 3, l: 9 })
  }
  // Green
  if (newColor.h >= 85 && newColor.h < 165) {
    return shiftHslColor(newColor, { h: 16, l: 14 })
  }
  // Blue
  if (newColor.h >= 165 && newColor.h < 285) {
    return shiftHslColor(newColor, { h: -15, s: 3, l: 2 })
  }

  throw new Error('This method should handle all colors. But it didn\'t for some reason.')
}

export const getGradientBackground = (color: string) => {
  const colorLeft = shiftGradientColor(color)
  const colorRight = ColorTranslator.toHSL(color)

  return `linear-gradient(to right, ${colorLeft}, ${colorRight})`
}
