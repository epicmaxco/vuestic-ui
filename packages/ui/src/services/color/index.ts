export * from './types'
export * from './config/default'
export * from './utils'
