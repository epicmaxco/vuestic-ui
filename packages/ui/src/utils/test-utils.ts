/**
 * @deprecated
 *  use vitest each
 * @param one
 * @param args
 */
export const fromTable = (one: any, ...args: any[]) => {
  const rows = one[0].split('|').map((i: any) => i.trim())
  let result: any[] = []
  while (args.length) {
    result = [...result, args.splice(0, rows.length)]
  }
  return result
}
