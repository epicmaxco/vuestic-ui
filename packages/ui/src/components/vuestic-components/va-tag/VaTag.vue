<template>
  <component
    :is="tagComputed"
    v-if="valueComputed"
    :href="href"
    :target="target"
    :to="to"
    :replace="replace"
    :exact="exact"
    :active-class="activeClass"
    :exact-active-class="exactActiveClass"
    class="va-tag"
    :class="computedClass"
    :style="computedStyle"
    :tabindex="indexComputed"
    @mouseenter="updateHoverState(true)"
    @mouseleave="updateHoverState(false)"
    @focus="updateFocusState(true)"
    @blur="updateFocusState(false)"
  >
    <va-icon
      v-if="c_icon"
      class="va-tag__icon"
      :name="icon"
      :size="iconSize"
    />
    <span class="va-tag__content">
      <slot></slot>
    </span>
    <va-icon
      v-if="c_closeable"
      class="va-tag__close-icon"
      name="close"
      :size="iconSize"
      @click.stop="close()"
    />
  </component>
</template>

<script lang="ts">
import { Component, Mixins, Watch } from 'vue-property-decorator'

import VaIcon from '../va-icon/VaIcon.vue'

import {
  getBoxShadowColor,
  getHoverColor,
  getFocusColor,
} from '../../../services/color-functions'
import { ColorThemeMixin } from '../../../services/ColorThemePlugin'
import { RouterLinkMixin } from '../../vuestic-mixins/RouterLinkMixin/RouterLinkMixin'
import { StatefulMixin } from '../../vuestic-mixins/StatefulMixin/StatefulMixin'
import { KeyboardOnlyFocusMixin } from '../../vuestic-mixins/KeyboardOnlyFocusMixin/KeyboardOnlyFocusMixin'
import { makeContextablePropsMixin } from '../../context-test/context-provide/ContextPlugin'

const TagPropsMixin = makeContextablePropsMixin({
  value: { type: Boolean, default: true },
  closeable: { type: Boolean, default: false },
  color: { type: String, default: '' },
  outline: { type: Boolean, default: false },
  flat: { type: Boolean, default: false },
  icon: { type: String, default: '' },
  disabled: { type: Boolean, default: false },
  square: { type: Boolean, default: false },
  tag: { type: String, default: 'span' },
  size: {
    type: String,
    default: 'medium',
    validator: (value: string) => {
      return ['medium', 'small', 'large'].includes(value)
    },
  },
})

@Component({
  name: 'VaTag',
  components: { VaIcon },
})
export default class VaTag extends Mixins(
  KeyboardOnlyFocusMixin,
  RouterLinkMixin,
  StatefulMixin,
  ColorThemeMixin,
  TagPropsMixin,
) {
  hoverState = false
  focusState = false

  @Watch('hoverState')
  onHoverChange (value: boolean) {
    this.updateFocusState(value)
    this.updateHoverState(value)
  }

  get iconSize () {
    const size: any = {
      small: '0.875rem',
      medium: '1rem',
      large: '1.25rem',
    }
    return size[this.c_size]
  }

  get indexComputed () {
    return this.c_disabled ? -1 : 0
  }

  get computedClass () {
    return {
      'va-tag--small': this.c_size === 'small',
      'va-tag--large': this.c_size === 'large',
      'va-tag--square': this.c_square,
      'va-tag--disabled': this.c_disabled,
    }
  }

  get shadowStyle () {
    if (this.c_flat || this.c_outline) {
      return
    }
    return '0 0.125rem 0.19rem 0 ' + getBoxShadowColor(this.colorComputed)
  }

  get computedStyle () {
    const computedStyle: any = {
      color: '',
      borderColor: '',
      background: '',
      boxShadow: '',
    }

    if (this.focusState) {
      if (this.c_outline || this.c_flat) {
        computedStyle.color = this.colorComputed
        computedStyle.borderColor = this.c_outline ? this.colorComputed : ''
        computedStyle.background = getFocusColor(this.colorComputed)
      }
    } else if (this.hoverState) {
      if (this.c_outline || this.c_flat) {
        computedStyle.color = this.colorComputed
        computedStyle.borderColor = this.c_outline ? this.colorComputed : ''
        computedStyle.background = getHoverColor(this.colorComputed)
      } else {
        computedStyle.boxShadow = this.shadowStyle
      }
    } else {
      computedStyle.color = this.c_flat || this.c_outline ? this.colorComputed : ''
      computedStyle.borderColor = this.c_outline ? this.colorComputed : ''
      computedStyle.boxShadow = !this.c_disabled && this.shadowStyle
    }

    if (!this.c_outline && !this.c_flat) {
      computedStyle.background = this.colorComputed
    }

    return computedStyle
  }

  updateHoverState (isHover: boolean) {
    this.hoverState = isHover
  }

  updateFocusState (isFocus: boolean) {
    this.focusState = isFocus
    if (isFocus) {
      this.isKeyboardFocused = isFocus
      this.KeyboardOnlyFocusMixin_onFocus()
      this.$emit('focus')
    }
  }

  close () {
    if (!this.c_disabled) {
      this.valueComputed = false
    }
  }
}
</script>

<style lang="scss">
@import "../../vuestic-sass/resources/resources";

$tag-font-size-nrm: 1rem !default;
$tag-font-size-sm: 0.875rem !default;
$tag-font-size-lg: 1.25rem !default;

.va-tag {
  display: inline-flex;
  border: 0.125rem solid transparent;
  position: relative;
  border-radius: 2rem;
  width: auto;
  height: auto;
  min-width: initial;
  min-height: initial;

  /* margin: 0 0.1rem; */
  padding: 0 0.3rem;
  color: $white;
  cursor: default;
  align-items: center;
  font-size: $tag-font-size-nrm;

  &:hover {
    opacity: 0.85;
  }

  &__content {
    display: flex;
    align-items: center;
    padding: 0 0.3rem;
    line-height: 1.6;
  }

  &__close-icon {
    cursor: pointer;

    /* z-index: 500; */

    @at-root {
      .va-tag--disabled {
        .va-tag__close-icon {
          cursor: default !important;
        }
      }
    }
  }

  &--square {
    border-radius: 0.2rem;
  }

  &--small {
    height: 1.5rem;
    font-size: $tag-font-size-sm;
  }

  &--large {
    height: 2.5rem;
    font-size: $tag-font-size-lg;
  }

  &.va-tag--disabled {
    @include va-disabled;
  }
}
</style>
