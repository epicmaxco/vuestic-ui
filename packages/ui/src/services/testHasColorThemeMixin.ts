import { shallowMount, Wrapper } from '@vue/test-utils'
import { ColorThemeMixin } from './ColorThemePlugin'
import { testIsContextableComponent } from '../components/context-test/context-provide/testIsContextableComponent'

export function testHasColorThemeMixin<T extends ColorThemeMixin> (componentOptions: T) {
  const wrapper: Wrapper<T> = shallowMount(componentOptions)

  // Test mixin applied
  if (!(wrapper.vm as any).hasColorThemeMixin) {
    throw new Error('ColorThemeMixin is not added')
  }

  // Test is contextable [troubles with icon context]
  // const props = {
  //   color: '',
  // }
  // expect(() => testIsContextableComponent(componentOptions, props)).not.toThrow()
}
