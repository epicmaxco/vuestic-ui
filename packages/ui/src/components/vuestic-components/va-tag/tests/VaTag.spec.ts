import { mount, shallowMount } from '@vue/test-utils'
import VaTag from '../VaTag.vue'
import { testHasColorThemeMixin } from '../../../../services/testHasColorThemeMixin'
import { ColorThemeMixin } from '../../../../services/ColorThemePlugin'

describe('VaTag', () => {
  it('should render without an error', () => {
    const wrapper = mount(VaTag)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
  /* 2. Button sizes */

  it('large tag', () => {
    const wrapper = shallowMount(VaTag, {
      propsData: { size: 'large' },
    })
    expect(wrapper.find(VaTag).classes()).toContain('va-tag--large')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('small tag', () => {
    const wrapper = shallowMount(VaTag, {
      propsData: { size: 'small' },
    })
    expect(wrapper.find(VaTag).classes()).toContain('va-tag--small')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be a link', () => {
    const wrapper = shallowMount(VaTag, {
      propsData: {
        href: '/',
      },
    })
    expect(wrapper.is('a')).toBe(true)
  })
  it('has ColorThemeMixin', () => {
    expect(() =>
      testHasColorThemeMixin((VaTag as unknown) as ColorThemeMixin),
    ).not.toThrow()
  })
})
