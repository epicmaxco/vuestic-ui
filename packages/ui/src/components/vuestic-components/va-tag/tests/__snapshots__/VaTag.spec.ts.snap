// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaTag large tag 1`] = `
<span tabindex="0" class="va-tag va-tag--large" style="background: rgb(0, 0, 0); box-shadow: 0 0.125rem 0.19rem 0 rgba(0,0,0,0.4);"><!----> <span class="va-tag__content"></span>
<!----></span>
`;

exports[`VaTag small tag 1`] = `
<span tabindex="0" class="va-tag va-tag--small" style="background: rgb(0, 0, 0); box-shadow: 0 0.125rem 0.19rem 0 rgba(0,0,0,0.4);"><!----> <span class="va-tag__content"></span>
<!----></span>
`;
